File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/RequestResultTest.java
package com.snowplowanalytics.snowplow.tracker;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertEquals;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.snowplowanalytics.snowplow.network.RequestResult;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

@RunWith(AndroidJUnit4.class)
public class RequestResultTest {

    @Test
    public void testSuccessfulRequest() {
        RequestResult result = new RequestResult(200, false, Arrays.asList(100L));
        assertTrue(result.isSuccessful());
        assertFalse(result.isOversize());
        assertFalse(result.shouldRetry(new HashMap<>()));
        assertEquals(result.getEventIds(), Arrays.asList(100L));
    }

    @Test
    public void testFailedRequest() {
        RequestResult result = new RequestResult(500, false, new ArrayList<>());
        assertFalse(result.isSuccessful());
        assertTrue(result.shouldRetry(new HashMap<>()));
    }

    @Test
    public void testOversizedFailedRequest() {
        RequestResult result = new RequestResult(500, true, new ArrayList<>());
        assertFalse(result.isSuccessful());
        assertFalse(result.shouldRetry(new HashMap<>()));
    }

    @Test
    public void testFailedRequestWithNoRetryStatus() {
        RequestResult result = new RequestResult(403, false, new ArrayList<>());
        assertFalse(result.isSuccessful());
        assertFalse(result.shouldRetry(new HashMap<>()));
    }

    @Test
    public void testFailedRequestWithCustomRetryRules() {
        Map<Integer, Boolean> customRetry = new HashMap<>();
        customRetry.put(403, true);
        customRetry.put(500, false);

        RequestResult result = new RequestResult(403, false, new ArrayList<>());
        assertFalse(result.isSuccessful());
        assertTrue(result.shouldRetry(customRetry));

        result = new RequestResult(500, false, new ArrayList<>());
        assertFalse(result.isSuccessful());
        assertFalse(result.shouldRetry(customRetry));
    }

}

File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/RequestResultTest.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.tracker

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.snowplowanalytics.snowplow.network.RequestResult
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

@RunWith(AndroidJUnit4::class)
class RequestResultTest {
    @Test
    fun testSuccessfulRequest() {
        val result = RequestResult(200, false, listOf(100L))
        Assert.assertTrue(result.isSuccessful)
        Assert.assertFalse(result.oversize)
        Assert.assertFalse(result.shouldRetry(HashMap()))
        Assert.assertEquals(result.eventIds, listOf(100L))
    }

    @Test
    fun testFailedRequest() {
        val result = RequestResult(500, false, ArrayList())
        Assert.assertFalse(result.isSuccessful)
        Assert.assertTrue(result.shouldRetry(HashMap()))
    }

    @Test
    fun testOversizedFailedRequest() {
        val result = RequestResult(500, true, ArrayList())
        Assert.assertFalse(result.isSuccessful)
        Assert.assertFalse(result.shouldRetry(HashMap()))
    }

    @Test
    fun testFailedRequestWithNoRetryStatus() {
        val result = RequestResult(403, false, ArrayList())
        Assert.assertFalse(result.isSuccessful)
        Assert.assertFalse(result.shouldRetry(HashMap()))
    }

    @Test
    fun testFailedRequestWithCustomRetryRules() {
        val customRetry: MutableMap<Int, Boolean> = HashMap()
        customRetry[403] = true
        customRetry[500] = false
        var result = RequestResult(403, false, ArrayList())
        Assert.assertFalse(result.isSuccessful)
        Assert.assertTrue(result.shouldRetry(customRetry))
        result = RequestResult(500, false, ArrayList())
        Assert.assertFalse(result.isSuccessful)
        Assert.assertFalse(result.shouldRetry(customRetry))
    }
}
